### PR TITLE
Fix rings in dyn waves at low framerates

### PR DESCRIPTION
**Pull request by huwb**
**Status**: Ready for review

Even though the ripple sim input is applied at 60hz+, it only takes the SphereWaterInteraction position in LateUpdate which could be 30hz. So its a framerate-dependent issue and i was not seeing it previously because my framerate was too high.

This PR corrects the dynamic wave input to account for the simulation time.

An alternative approach would be to schedule the sim update from FixedUpdate, but this would execute before the water system is up to date, and it would only work at the fixed delta time.

Other fixes found along the way:

* Removed unnecessary angle clamp applied to SWI object
* Put SWI shader in Resources folder

**Testing**: *BoatProbes* in *boat.unity* is hacked to move fast and there is a *Sleeper* component on the SWI object. The SWI component has a *Do Vel Correction* param to enable/disable this change. Red debug lines are drawn at each reconstructed location for visual inspection.

### CODE REVIEW

**Review by daleeidd — COMMENTED** [review]
Nice. Looks a lot better. The effect is different (quite a bit stronger) with the sleeper enabled. Is this expected? I'm not expecting interpolation to be perfect but just thought I check.

Screenshot (358)
Sleeper disabled

Screenshot (357)
Sleeper enabled

**Comment by daleeidd** [comment]
@huwb We should push this forward so it can be included in this release. I have to move the shader to a Resources folder as per #971.

**Comment by huwb** [comment]
Ok will aim to look at this tomorrow

File: crest/Assets/Crest/Crest/Scripts/Interaction/SphereWaterInteraction.cs [hunk 1]
```diff
@@ -169,15 +167,13 @@ Vector3 LateUpdateComputeVelRelativeToWater(OceanRenderer ocean)
             else if (speedKmh > _maxSpeed)
             {
                 // limit speed to max
-                vel *= _maxSpeed / speedKmh;
+                _velocity *= _maxSpeed / speedKmh;
```
[the remark this answers is not in the document]
Comment (huwb) [reply crest/Assets/Crest/Crest/Scripts/Interaction/SphereWaterInteraction.cs]: was realllly tempted to just remove the clamped vel. but i ended up going with your suggestion - store clamped vel separately. nice spot again.

**Comment by huwb** [comment]
> The effect is different (quite a bit stronger) with the sleeper enabled

So at the time i think i remember seeing this but didnt know how to fix it.

Testing now, i dont seem to be able to repro this issue. The wake set up in boat.unity on this stream seems to look identical with sleeper=0 and sleeper=50

Could i possible get a retest?

**Comment by huwb** [comment]
And increasing the simulation frequency makes the effect smoother which is great and might help some cases.

**Comment by huwb** [comment]
Ok besides that question above i think this is ready for review

**Comment by daleeidd** [comment]
> > The effect is different (quite a bit stronger) with the sleeper enabled
> 
> So at the time i think i remember seeing this but didnt know how to fix it.
> 
> Testing now, i dont seem to be able to repro this issue. The wake set up in boat.unity on this stream seems to look identical with sleeper=0 and sleeper=50
> 
> Could i possible get a retest?

Sure. It is more noticeable in the first LOD when the camera is close to the surface.

Screenshot (379)_LI
Screenshot (378)_LI

File: docs/about/history.rst [hunk 2]
```diff
@@ -64,6 +64,7 @@ Fixed
    -  Fix *Allow Null Light* option on *Sim Settings Shadows* not working.
    -  Fix ocean tiles not reverting to *Ocean Renderer > Material* if *Water Body > Override Material* was used and *Water Body* was disabled or removed.
    -  Fix underwater rendering when the camera's culling mask excludes the *Ocean Renderer > Layer*.
+   -  Fix visible "rings" from fast moving dynamic wave sim inputs. Simulation frequency can be increased to improve result further, at the cost of more simulation steps per frame.
```
Comment (daleeidd) [inline docs/about/history.rst:67]: - needs a new line after period for style
- "dynamic wave sim inputs" shouldn't this only be for sphere water interaction?
- an entry for fixing the shader being moved which fixes builds

**Review by daleeidd — APPROVED** [review]
Just minor documentation changes. Otherwise approved. Nice work fixing this.

File: crest/Assets/Crest/Crest/Scripts/Interaction/SphereWaterInteraction.cs [hunk 3]
```diff
@@ -137,47 +145,43 @@ void LateUpdate()
...
 
         // Velocity of the sphere, relative to the water. Computes on the fly, discards if teleport detected.
-        Vector3 LateUpdateComputeVelRelativeToWater(OceanRenderer ocean)
+        void LateUpdateComputeVel(OceanRenderer ocean)
         {
-            Vector3 vel;
-
-            // feed in water velocity
-            vel = (transform.position - _posLast) / ocean.DeltaTimeDynamics;
+            // Compue vel using finite difference
```
[the remark this answers is not in the document]
Comment (huwb) [reply crest/Assets/Crest/Crest/Scripts/Interaction/SphereWaterInteraction.cs:150]: Done

File: docs/user/ocean-simulation.rst [hunk 4]
```diff
@@ -115,6 +115,9 @@ For an example of usages in boats, search for GameObjects with "InteractionSpher
 
 The dynamic wave simulation is added on top of the animated FFT waves to give the final shape.
 
+The dynamic wave simulation is not suitable for use further than approximately 10km from the origin.
+At this kind of distance the stability of the simulation can be compromised.
+Use the *FloatingOrigin*  component to avoid travelling far distances from the world origin.
```
[the remark this answers is not in the document]
Comment (huwb) [reply docs/user/ocean-simulation.rst:120]: Done

File: docs/user/ocean-simulation.rst [hunk 5]
```diff
@@ -123,6 +126,8 @@ Simulation Settings
 
 All of the settings below refer to the *Dynamic Wave Sim Settings* asset.
 
+The key settings that impact stability of hte simulation are the **Damping** and **Courant Number** settings described below.
```
[the remark this answers is not in the document]
Comment (huwb) [reply docs/user/ocean-simulation.rst]: Done

File: docs/about/history.rst (hunk 2 above)
Comment (huwb) [reply docs/about/history.rst]: Done

**Comment by huwb** [comment]
Oops sorry for all those issues

> It is more noticeable in the first LOD when the camera is close to the surface

Ok I'm not sure how to fix it so I think I'll defer it for now.

**Comment by nitroman123** [comment]
Unfortunately they are still there tested today... 
Screenshot 2022-01-08 144323

**Pull request merged**